File: probes/memory/memory-tracker/memory_tracker.c
/*
 * Memory Tracker eBPF Probe
 * Monitors memory allocation, deallocation, and usage patterns
 * 
 * This probe tracks:
 * - Memory allocations and deallocations
 * - Page faults and memory pressure
 * - Process memory usage (RSS, VSZ, heap)
 * - System-wide memory statistics
 * - Memory leaks and fragmentation
 */

#include <vmlinux.h>
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>
#include <bpf/bpf_core_read.h>

#define MAX_ENTRIES 10240
#define MAX_STACK_DEPTH 20
#define TASK_COMM_LEN 16

/* Memory allocation event types */
enum alloc_type {
    ALLOC_MALLOC = 1,
    ALLOC_CALLOC,
    ALLOC_REALLOC,
    ALLOC_FREE,
    ALLOC_MMAP,
    ALLOC_MUNMAP,
    ALLOC_BRK,
    ALLOC_PAGE,
};

/* Data structures */
struct memory_event {
    __u64 timestamp;
    __u32 pid;
    __u32 tid;
    __u64 addr;
    __u64 size;
    __u64 old_addr;  // for realloc
    __u32 type;      // enum alloc_type
    __u32 flags;
    __u64 stack_id;
    char comm[TASK_COMM_LEN];
};

struct process_memory {
    __u64 total_allocated;
    __u64 total_freed;
    __u64 current_usage;
    __u64 peak_usage;
    __u64 allocation_count;
    __u64 free_count;
    __u64 page_faults;
    __u64 major_faults;
    __u64 rss_pages;
    __u64 vmem_pages;
};

struct system_memory {
    __u64 total_memory;
    __u64 free_memory;
    __u64 available_memory;
    __u64 cached_memory;
    __u64 buffer_memory;
    __u64 slab_memory;
    __u64 page_cache_size;
    __u32 memory_pressure;
};

struct allocation_info {
    __u64 size;
    __u64 timestamp;
    __u64 stack_id;
    __u32 pid;
};

/* BPF Maps */
struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, MAX_ENTRIES);
    __type(key, __u32); // PID
    __type(value, struct process_memory);
} process_memory_map SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, MAX_ENTRIES * 4);
    __type(key, __u64); // Address
    __type(value, struct allocation_info);
} allocation_map SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_ARRAY);
    __uint(max_entries, 1);
    __type(key, __u32);
    __type(value, struct system_memory);
} system_memory_map SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_STACK_TRACE);
    __uint(max_entries, 1024);
    __uint(key_size, sizeof(__u32));
    __uint(value_size, MAX_STACK_DEPTH * sizeof(__u64));
} stack_traces SEC(".maps");

struct {
    __uint(type, BPF_MAP_TYPE_RINGBUF);
    __uint(max_entries, 256 * 1024);
} events SEC(".maps");

/* Configuration */
struct {
    __uint(type, BPF_MAP_TYPE_ARRAY);
    __uint(max_entries, 4);
    __type(key, __u32);
    __type(value, __u32);
} config_map SEC(".maps");

/* Helper function to send memory event to userspace */
static __always_inline void send_memory_event(__u32 pid, __u64 addr, 
                                             __u64 size, __u32 type,
                                             __u64 old_addr) {
    struct memory_event *event;
    
    event = bpf_ringbuf_reserve(&events, sizeof(*event), 0);
    if (!event)
        return;
    
    event->timestamp = bpf_ktime_get_ns();
    event->pid = pid;
    event->tid = bpf_get_current_pid_tgid();
    event->addr = addr;
    event->size = size;
    event->old_addr = old_addr;
    event->type = type;
    event->flags = 0;
    
    // Capture stack trace
    event->stack_id = bpf_get_stackid(bpf_get_current_task(), &stack_traces, 
                                     BPF_F_USER_STACK);
    
    // Get process name
    bpf_get_current_comm(&event->comm, sizeof(event->comm));
    
    bpf_ringbuf_submit(event, 0);
}

/* Helper function to update process memory statistics */
static __always_inline void update_process_memory(__u32 pid, __s64 size_delta,
                                                 __u32 is_allocation) {
    struct process_memory *mem = bpf_map_lookup_elem(&process_memory_map, &pid);
    if (!mem) {
        struct process_memory new_mem = {};
        bpf_map_update_elem(&process_memory_map, &pid, &new_mem, BPF_ANY);
        mem = bpf_map_lookup_elem(&process_memory_map, &pid);
        if (!mem) return;
    }
    
    if (is_allocation) {
        mem->total_allocated += size_delta;
        mem->allocation_count++;
        mem->current_usage += size_delta;
        if (mem->current_usage > mem->peak_usage) {
            mem->peak_usage = mem->current_usage;
        }
    } else {
        mem->total_freed += size_delta;
        mem->free_count++;
        if (mem->current_usage >= size_delta) {
            mem->current_usage -= size_delta;
        }
    }
}

/* Trace malloc calls */
SEC("uprobe/malloc")
int trace_malloc(struct pt_regs *ctx) {
    __u64 size = PT_REGS_PARM1(ctx);
    __u32 pid = bpf_get_current_pid_tgid() >> 32;
    
    if (pid == 0 || size == 0)
        return 0;
    
    send_memory_event(pid, 0, size, ALLOC_MALLOC, 0);
    return 0;
}

SEC("uretprobe/malloc")
int trace_malloc_ret(struct pt_regs *ctx) {
    __u64 addr = PT_REGS_RC(ctx);
    __u32 pid = bpf_get_current_pid_tgid() >> 32;
    
    if (pid == 0 || addr == 0)
        return 0;
    
    // We need to correlate this with the size from the entry probe
    // For now, we'll store the allocation in a temporary map
    return 0;
}

/* Trace free calls */
SEC("uprobe/free")
int trace_free(struct pt_regs *ctx) {
    __u64 addr = PT_REGS_PARM1(ctx);
    __u32 pid = bpf_get_current_pid_tgid() >> 32;
    
    if (pid == 0 || addr == 0)
        return 0;
    
    // Look up allocation info
    struct allocation_info *info = bpf_map_lookup_elem(&allocation_map, &addr);
    __u64 size = 0;
    if (info) {
        size = info->size;
        bpf_map_delete_elem(&allocation_map, &addr);
        update_process_memory(pid, size, 0);
    }
    
    send_memory_event(pid, addr, size, ALLOC_FREE, 0);
    return 0;
}

/* Trace mmap calls */
SEC("tp/syscalls/sys_enter_mmap")
int trace_mmap_enter(struct trace_event_raw_sys_enter *ctx) {
    __u64 size = ((__u64)ctx->args[1]);
    __u32 pid = bpf_get_current_pid_tgid() >> 32;
    
    if (pid == 0 || size == 0)
        return 0;
    
    send_memory_event(pid, 0, size, ALLOC_MMAP, 0);
    return 0;
}

SEC("tp/syscalls/sys_exit_mmap")
int trace_mmap_exit(struct trace_event_raw_sys_exit *ctx) {
    __u64 addr = ctx->ret;
    __u32 pid = bpf_get_current_pid_tgid() >> 32;
    
    if (pid == 0 || (__s64)addr < 0)
        return 0;
    
    // Store allocation info for future munmap
    struct allocation_info info = {};
    info.timestamp = bpf_ktime_get_ns();
    info.pid = pid;
    // info.size would need to be retrieved from entry
    
    bpf_map_update_elem(&allocation_map, &addr, &info, BPF_ANY);
    return 0;
}

/* Trace munmap calls */
SEC("tp/syscalls/sys_enter_munmap")
int trace_munmap(struct trace_event_raw_sys_enter *ctx) {
    __u64 addr = ctx->args[0];
    __u64 size = ctx->args[1];
    __u32 pid = bpf_get_current_pid_tgid() >> 32;
    
    if (pid == 0 || addr == 0)
        return 0;
    
    struct allocation_info *info = bpf_map_lookup_elem(&allocation_map, &addr);
    if (info) {
        bpf_map_delete_elem(&allocation_map, &addr);
        update_process_memory(pid, size, 0);
    }
    
    send_memory_event(pid, addr, size, ALLOC_MUNMAP, 0);
    return 0;
}

/* Trace brk syscall */
SEC("tp/syscalls/sys_enter_brk")
int trace_brk(struct trace_event_raw_sys_enter *ctx) {
    __u64 addr = ctx->args[0];
    __u32 pid = bpf_get_current_pid_tgid() >> 32;
    
    if (pid == 0)
        return 0;
    
    send_memory_event(pid, addr, 0, ALLOC_BRK, 0);
    return 0;
}

/* Trace page faults */
SEC("tp/exceptions/page_fault_user")
int trace_page_fault(struct trace_event_raw_page_fault_user *ctx) {
    __u32 pid = bpf_get_current_pid_tgid() >> 32;
    __u64 address = ctx->address;
    __u32 error_code = ctx->error_code;
    
    if (pid == 0)
        return 0;
    
    struct process_memory *mem = bpf_map_lookup_elem(&process_memory_map, &pid);
    if (!mem) {
        struct process_memory new_mem = {};
        bpf_map_update_elem(&process_memory_map, &pid, &new_mem, BPF_ANY);
        mem = bpf_map_lookup_elem(&process_memory_map, &pid);
        if (!mem) return 0;
    }
    
    mem->page_faults++;
    
    // Check if this is a major fault (page not in memory)
    if (error_code & 0x4) { // Major fault flag
        mem->major_faults++;
    }
    
    send_memory_event(pid, address, 4096, ALLOC_PAGE, 0);
    return 0;
}

/* Monitor memory pressure events */
SEC("tp/vmscan/mm_vmscan_wakeup_kswapd")
int trace_memory_pressure(void *ctx) {
    __u32 key = 0;
    struct system_memory *sys_mem = bpf_map_lookup_elem(&system_memory_map, &key);
    if (sys_mem) {
        sys_mem->memory_pressure++;
    }
    return 0;
}

/* Trace OOM killer events */
SEC("tp/oom/mark_victim")
int trace_oom_victim(struct trace_event_raw_mark_victim *ctx) {
    __u32 pid = ctx->pid;
    
    // Send OOM event
    send_memory_event(pid, 0, 0, 0xFF, 0); // Special type for OOM
    return 0;
}

/* Sample memory statistics periodically */
SEC("perf_event")
int sample_memory_stats(struct bpf_perf_event_data *ctx) {
    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
    __u32 pid = bpf_get_current_pid_tgid() >> 32;
    
    if (pid == 0)
        return 0;
    
    // Read memory statistics from task struct
    struct mm_struct *mm;
    BPF_CORE_READ_INTO(&mm, task, mm);
    
    if (!mm)
        return 0;
    
    struct process_memory *mem = bpf_map_lookup_elem(&process_memory_map, &pid);
    if (!mem) {
        struct process_memory new_mem = {};
        bpf_map_update_elem(&process_memory_map, &pid, &new_mem, BPF_ANY);
        mem = bpf_map_lookup_elem(&process_memory_map, &pid);
        if (!mem) return 0;
    }
    
    // Update RSS and virtual memory statistics
    __u64 rss_pages, vmem_pages;
    BPF_CORE_READ_INTO(&rss_pages, mm, rss_stat.count[0].counter);
    BPF_CORE_READ_INTO(&vmem_pages, mm, total_vm);
    
    mem->rss_pages = rss_pages;
    mem->vmem_pages = vmem_pages;
    
    return 0;
}

/* Kprobe for detailed allocation tracking */
SEC("kprobe/__alloc_pages")
int BPF_KPROBE(__alloc_pages, gfp_t gfp_mask, unsigned int order) {
    __u32 pid = bpf_get_current_pid_tgid() >> 32;
    __u64 size = (1ULL << order) * 4096; // Pages to bytes
    
    if (pid == 0)
        return 0;
    
    update_process_memory(pid, size, 1);
    send_memory_event(pid, 0, size, ALLOC_PAGE, 0);
    return 0;
}

SEC("kprobe/__free_pages")
int BPF_KPROBE(__free_pages, struct page *page, unsigned int order) {
    __u32 pid = bpf_get_current_pid_tgid() >> 32;
    __u64 size = (1ULL << order) * 4096; // Pages to bytes
    
    if (pid == 0)
        return 0;
    
    update_process_memory(pid, size, 0);
    return 0;
}

char LICENSE[] SEC("license") = "GPL";

File: probes/memory/memory-tracker/memory_tracker.go
// Memory Tracker Userspace Agent
// Collects and processes memory allocation data from eBPF probe

package main

import (
    "bytes"
    "context"
    "encoding/binary"
    "fmt"
    "log"
    "os"
    "os/signal"
    "sort"
    "syscall"
    "time"
    "unsafe"

    "github.com/cilium/ebpf"
    "github.com/cilium/ebpf/link"
    "github.com/cilium/ebpf/ringbuf"
    "github.com/cilium/ebpf/rlimit"
)

// Memory allocation types
const (
    AllocMalloc = 1
    AllocCalloc = 2
    AllocRealloc = 3
    AllocFree = 4
    AllocMmap = 5
    AllocMunmap = 6
    AllocBrk = 7
    AllocPage = 8
    AllocOOM = 0xFF
)

var allocTypeNames = map[uint32]string{
    AllocMalloc:  "malloc",
    AllocCalloc:  "calloc", 
    AllocRealloc: "realloc",
    AllocFree:    "free",
    AllocMmap:    "mmap",
    AllocMunmap:  "munmap",
    AllocBrk:     "brk",
    AllocPage:    "page",
    AllocOOM:     "oom",
}

// Data structures matching eBPF program
type MemoryEvent struct {
    Timestamp uint64
    PID       uint32
    TID       uint32
    Addr      uint64
    Size      uint64
    OldAddr   uint64
    Type      uint32
    Flags     uint32
    StackID   uint64
    Comm      [16]int8
}

type ProcessMemory struct {
    TotalAllocated  uint64
    TotalFreed      uint64
    CurrentUsage    uint64
    PeakUsage       uint64
    AllocationCount uint64
    FreeCount       uint64
    PageFaults      uint64
    MajorFaults     uint64
    RSSPages        uint64
    VMemPages       uint64
}

type SystemMemory struct {
    TotalMemory     uint64
    FreeMemory      uint64
    AvailableMemory uint64
    CachedMemory    uint64
    BufferMemory    uint64
    SlabMemory      uint64
    PageCacheSize   uint64
    MemoryPressure  uint32
}

type AllocationInfo struct {
    Size      uint64
    Timestamp uint64
    StackID   uint64
    PID       uint32
}

type MemoryTracker struct {
    spec        *ebpf.CollectionSpec
    coll        *ebpf.Collection
    eventReader *ringbuf.Reader
    links       []link.Link
    
    // Statistics
    totalEvents       uint64
    allocationEvents  uint64
    freeEvents        uint64
    pageEvents        uint64
    oomEvents         uint64
    processStats      map[uint32]*ProcessMemory
    leaks             map[uint64]*AllocationInfo
    startTime         time.Time
}

func NewMemoryTracker() (*MemoryTracker, error) {
    if err := rlimit.RemoveMemlock(); err != nil {
        return nil, fmt.Errorf("failed to remove memlock: %v", err)
    }

    tracker := &MemoryTracker{
        processStats: make(map[uint32]*ProcessMemory),
        leaks:        make(map[uint64]*AllocationInfo),
        startTime:    time.Now(),
    }

    return tracker, nil
}

func (mt *MemoryTracker) Load() error {
    spec, err := ebpf.LoadCollectionSpec("memory_tracker.o")
    if err != nil {
        return fmt.Errorf("failed to load eBPF spec: %v", err)
    }
    mt.spec = spec

    coll, err := ebpf.NewCollection(spec)
    if err != nil {
        return fmt.Errorf("failed to create eBPF collection: %v", err)
    }
    mt.coll = coll

    // Create event reader
    reader, err := ringbuf.NewReader(coll.Maps["events"])
    if err != nil {
        return fmt.Errorf("failed to create ring buffer reader: %v", err)
    }
    mt.eventReader = reader

    return nil
}

func (mt *MemoryTracker) Attach() error {
    // Attach tracepoints
    tracepoints := []struct {
        group string
        name  string
        prog  string
    }{
        {"syscalls", "sys_enter_mmap", "trace_mmap_enter"},
        {"syscalls", "sys_exit_mmap", "trace_mmap_exit"},
        {"syscalls", "sys_enter_munmap", "trace_munmap"},
        {"syscalls", "sys_enter_brk", "trace_brk"},
        {"exceptions", "page_fault_user", "trace_page_fault"},
        {"vmscan", "mm_vmscan_wakeup_kswapd", "trace_memory_pressure"},
        {"oom", "mark_victim", "trace_oom_victim"},
    }
    
    for _, tp := range tracepoints {
        l, err := link.Tracepoint(link.TracepointOptions{
            Group:   tp.group,
            Name:    tp.name,
            Program: mt.coll.Programs[tp.prog],
        })
        if err != nil {
            log.Printf("Warning: failed to attach tracepoint %s:%s: %v", tp.group, tp.name, err)
            continue
        }
        mt.links = append(mt.links, l)
    }

    // Attach kprobes for kernel allocation tracking
    kprobes := []struct {
        symbol string
        prog   string
    }{
        {"__alloc_pages", "__alloc_pages"},
        {"__free_pages", "__free_pages"},
    }
    
    for _, kp := range kprobes {
        l, err := link.Kprobe(link.KprobeOptions{
            Symbol:  kp.symbol,
            Program: mt.coll.Programs[kp.prog],
        })
        if err != nil {
            log.Printf("Warning: failed to attach kprobe %s: %v", kp.symbol, err)
            continue
        }
        mt.links = append(mt.links, l)
    }

    // Try to attach uprobes for malloc/free tracking
    // Note: This requires the binary path and may fail in some environments
    mt.attachUprobes()

    log.Printf("Attached %d eBPF programs", len(mt.links))
    return nil
}

func (mt *MemoryTracker) attachUprobes() {
    // Common libc paths to try
    libcPaths := []string{
        "/lib/x86_64-linux-gnu/libc.so.6",
        "/usr/lib/x86_64-linux-gnu/libc.so.6",
        "/lib64/libc.so.6",
        "/usr/lib64/libc.so.6",
    }
    
    functions := []string{"malloc", "free"}
    
    for _, libcPath := range libcPaths {
        if _, err := os.Stat(libcPath); err != nil {
            continue
        }
        
        for _, funcName := range functions {
            // Attach uprobe
            l, err := link.Uprobe(link.UprobeOptions{
                Path:    libcPath,
                Symbol:  funcName,
                Program: mt.coll.Programs["trace_"+funcName],
            })
            if err != nil {
                log.Printf("Warning: failed to attach uprobe %s:%s: %v", libcPath, funcName, err)
                continue
            }
            mt.links = append(mt.links, l)
            
            // Attach uretprobe for malloc
            if funcName == "malloc" {
                l, err := link.Uprobe(link.UprobeOptions{
                    Path:       libcPath,
                    Symbol:     funcName,
                    Program:    mt.coll.Programs["trace_malloc_ret"],
                    ReturnProbe: true,
                })
                if err != nil {
                    log.Printf("Warning: failed to attach uretprobe %s:%s: %v", libcPath, funcName, err)
                    continue
                }
                mt.links = append(mt.links, l)
            }
        }
        break // Use first available libc
    }
}

func (mt *MemoryTracker) processEvent(record ringbuf.Record) error {
    if len(record.RawSample) < int(unsafe.Sizeof(MemoryEvent{})) {
        return fmt.Errorf("invalid sample size")
    }

    var event MemoryEvent
    err := binary.Read(bytes.NewReader(record.RawSample), binary.LittleEndian, &event)
    if err != nil {
        return fmt.Errorf("failed to parse event: %v", err)
    }

    mt.totalEvents++
    
    // Convert C string to Go string
    comm := make([]byte, 0, 16)
    for _, c := range event.Comm {
        if c == 0 {
            break
        }
        comm = append(comm, byte(c))
    }
    
    // Update statistics based on event type
    switch event.Type {
    case AllocMalloc, AllocMmap, AllocBrk, AllocPage:
        mt.allocationEvents++
        mt.trackAllocation(event.PID, event.Addr, event.Size)
    case AllocFree, AllocMunmap:
        mt.freeEvents++
        mt.trackDeallocation(event.PID, event.Addr, event.Size)
    case AllocOOM:
        mt.oomEvents++
        log.Printf("OOM event detected for PID %d (%s)", event.PID, string(comm))
    }
    
    // Print interesting events
    if event.Size > 1024*1024 || event.Type == AllocOOM { // Large allocations or OOM
        typeName, ok := allocTypeNames[event.Type]
        if !ok {
            typeName = fmt.Sprintf("unknown(%d)", event.Type)
        }
        
        fmt.Printf("Memory Event: PID=%d, Type=%s, Addr=0x%x, Size=%d, Comm=%s\n",
            event.PID, typeName, event.Addr, event.Size, string(comm))
    }

    return nil
}

func (mt *MemoryTracker) trackAllocation(pid uint32, addr, size uint64) {
    if addr == 0 {
        return
    }
    
    // Track potential leaks
    mt.leaks[addr] = &AllocationInfo{
        Size:      size,
        Timestamp: time.Now().UnixNano(),
        PID:       pid,
    }
    
    // Update process statistics
    if _, exists := mt.processStats[pid]; !exists {
        mt.processStats[pid] = &ProcessMemory{}
    }
    
    stats := mt.processStats[pid]
    stats.TotalAllocated += size
    stats.AllocationCount++
    stats.CurrentUsage += size
    
    if stats.CurrentUsage > stats.PeakUsage {
        stats.PeakUsage = stats.CurrentUsage
    }
}

func (mt *MemoryTracker) trackDeallocation(pid uint32, addr, size uint64) {
    if addr == 0 {
        return
    }
    
    // Remove from leak tracking
    if _, exists := mt.leaks[addr]; exists {
        delete(mt.leaks, addr)
    }
    
    // Update process statistics
    if stats, exists := mt.processStats[pid]; exists {
        stats.TotalFreed += size
        stats.FreeCount++
        if stats.CurrentUsage >= size {
            stats.CurrentUsage -= size
        }
    }
}

func (mt *MemoryTracker) Run(ctx context.Context) error {
    fmt.Println("Starting memory tracker...")

    for {
        select {
        case <-ctx.Done():
            return ctx.Err()
        default:
            record, err := mt.eventReader.Read()
            if err != nil {
                if err == ringbuf.ErrClosed {
                    return nil
                }
                log.Printf("Error reading from ring buffer: %v", err)
                continue
            }

            if err := mt.processEvent(record); err != nil {
                log.Printf("Error processing event: %v", err)
            }
        }
    }
}

func (mt *MemoryTracker) PrintStats() {
    fmt.Printf("\n=== Memory Tracker Statistics ===\n")
    fmt.Printf("Runtime: %v\n", time.Since(mt.startTime))
    fmt.Printf("Total events: %d\n", mt.totalEvents)
    fmt.Printf("Allocation events: %d\n", mt.allocationEvents)
    fmt.Printf("Free events: %d\n", mt.freeEvents)
    fmt.Printf("Page fault events: %d\n", mt.pageEvents)
    fmt.Printf("OOM events: %d\n", mt.oomEvents)
    fmt.Printf("Tracked processes: %d\n", len(mt.processStats))
    fmt.Printf("Potential leaks: %d\n", len(mt.leaks))

    // Top memory consumers
    fmt.Printf("\nTop 10 memory consumers:\n")
    type processInfo struct {
        pid     uint32
        current uint64
        peak    uint64
        allocs  uint64
    }
    
    var processes []processInfo
    for pid, stats := range mt.processStats {
        processes = append(processes, processInfo{
            pid:     pid,
            current: stats.CurrentUsage,
            peak:    stats.PeakUsage,
            allocs:  stats.AllocationCount,
        })
    }
    
    sort.Slice(processes, func(i, j int) bool {
        return processes[i].current > processes[j].current
    })
    
    count := len(processes)
    if count > 10 {
        count = 10
    }
    
    for i := 0; i < count; i++ {
        p := processes[i]
        fmt.Printf("  PID %d: Current=%s, Peak=%s, Allocs=%d\n", 
            p.pid, formatBytes(p.current), formatBytes(p.peak), p.allocs)
    }
    
    // Memory leaks
    if len(mt.leaks) > 0 {
        fmt.Printf("\nPotential memory leaks (top 10):\n")
        type leakInfo struct {
            addr uint64
            size uint64
            age  time.Duration
            pid  uint32
        }
        
        var leaks []leakInfo
        now := time.Now().UnixNano()
        for addr, info := range mt.leaks {
            leaks = append(leaks, leakInfo{
                addr: addr,
                size: info.Size,
                age:  time.Duration(now - info.Timestamp),
                pid:  info.PID,
            })
        }
        
        sort.Slice(leaks, func(i, j int) bool {
            return leaks[i].size > leaks[j].size
        })
        
        leakCount := len(leaks)
        if leakCount > 10 {
            leakCount = 10
        }
        
        for i := 0; i < leakCount; i++ {
            l := leaks[i]
            fmt.Printf("  Addr=0x%x, Size=%s, Age=%v, PID=%d\n",
                l.addr, formatBytes(l.size), l.age.Truncate(time.Second), l.pid)
        }
    }
    
    // Read current memory statistics from maps
    mt.readMemoryMaps()
}

func (mt *MemoryTracker) readMemoryMaps() {
    processMap := mt.coll.Maps["process_memory_map"]
    
    fmt.Printf("\nProcess Memory Map (from eBPF):\n")
    var key uint32
    var stats ProcessMemory
    iter := processMap.Iterate()
    
    count := 0
    for iter.Next(&key, &stats) && count < 5 {
        fmt.Printf("  PID %d: Alloc=%s, Free=%s, Current=%s, Peak=%s\n",
            key, 
            formatBytes(stats.TotalAllocated),
            formatBytes(stats.TotalFreed),
            formatBytes(stats.CurrentUsage),
            formatBytes(stats.PeakUsage))
        count++
    }
}

func formatBytes(bytes uint64) string {
    const unit = 1024
    if bytes < unit {
        return fmt.Sprintf("%dB", bytes)
    }
    div, exp := int64(unit), 0
    for n := bytes / unit; n >= unit; n /= unit {
        div *= unit
        exp++
    }
    return fmt.Sprintf("%.1f%cB", float64(bytes)/float64(div), "KMGTPE"[exp])
}

func (mt *MemoryTracker) Close() error {
    if mt.eventReader != nil {
        mt.eventReader.Close()
    }

    for _, l := range mt.links {
        l.Close()
    }

    if mt.coll != nil {
        mt.coll.Close()
    }

    return nil
}

func main() {
    tracker, err := NewMemoryTracker()
    if err != nil {
        log.Fatalf("Failed to create memory tracker: %v", err)
    }
    defer tracker.Close()

    if err := tracker.Load(); err != nil {
        log.Fatalf("Failed to load eBPF program: %v", err)
    }

    if err := tracker.Attach(); err != nil {
        log.Fatalf("Failed to attach eBPF programs: %v", err)
    }

    // Handle interrupts gracefully
    ctx, cancel := context.WithCancel(context.Background())
    sigChan := make(chan os.Signal, 1)
    signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

    go func() {
        <-sigChan
        log.Println("Received interrupt signal, shutting down...")
        cancel()
    }()

    // Start stats printer goroutine
    go func() {
        ticker := time.NewTicker(15 * time.Second)
        defer ticker.Stop()
        
        for {
            select {
            case <-ctx.Done():
                return
            case <-ticker.C:
                tracker.PrintStats()
            }
        }
    }()

    // Run the tracker
    if err := tracker.Run(ctx); err != nil && err != context.Canceled {
        log.Fatalf("Memory tracker error: %v", err)
    }

    // Print final statistics
    tracker.PrintStats()
    log.Println("Memory tracker stopped")
}